### ISSUE TITLE
compaction failed: invalid_mutation_fragment_stream

### ISSUE THREAD

**Issue by chripede** [scylladb/scylla#9983]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 4.5.3-0.20211223.c8f14886d with build-id 9a5b504c51cbe8feb1217517d6977f7793b2971e
Cluster size: 4
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu

*Hardware details (for performance issues)*          Delete if unneeded
Platform (physical/VM/cloud instance type/docker): physical
Hardware: sockets=1 cores=4/6 hyperthreading=8/12 memory=64GB
Disks: (SSD/HDD, count) 2 ssd per node

We are unable to decomission and compact on some of our nodes because we get errors like these:

`Jan 28 13:32:20 api-scylla-yarey scylla:  [shard 0] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x60005a669b00 for sstable writer /var/lib/scylla/data/partnerapi/search_history-2742b7e01a9311eaa39d000000000003/md-60298-big-Data.db (partnerapi.search_history 2742b7e0-1a93-11ea-a39d-000000000003)] Unexpected token: previous 7223777569886450903, current 7223774232450905258): retrying
`

Looking at the few bugs I can find this issue should have been resolved a year ago.

**Comment by bhalevy** [comment]
Let's try to understand how this happened.

Did you ever run repair on this cluster with different number of shards on different nodes?
If so, what version were you running then?

To fix this issue scrub in segregate mode can help writing the out-of-order partitions into new sstables so they can be compacted.

Cc @denesb

**Comment by chripede** [comment]
We upgraded from <4 to 4.5.3
We then tried to replace a running node - that failed.
We then added a new node and tried decommissioning the old node - that failed.
We then tried running repair but that resulted in the above error.

We are currently scrubbing to see if that will do it.

**Comment by r4fek** [comment]
Same here. Except I did not run repair.

```
lut 11 15:27:12 am4-dbs1 scylla[28828]:  [shard 42] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62a00ab4db00 for sstable writer /var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/md-1343930-big-Data.db (sync.session_by_id 051e3940-94ad-11e8-93cf-000000000017)] Unexpected token: previous -4261495074737813676, current -4261527946879923112): retrying
```

**Comment by r4fek** [comment]
@chripede did you manage to fix the issue by scrubing the sstables? I tried that with no luck. I'm still unable to compact the tables and I get this error for multiple tables in the keyspace...

**Comment by chripede** [comment]
> @chripede did you manage to fix the issue by scrubing the sstables? I tried that with no luck. I'm still unable to compact the tables and I get this error for multiple tables in the keyspace...

No, in the end we had to remove the node by just shutting it down

**Comment by r4fek** [comment]
@bhalevy is there any other way to fix this? Unofortunately all my nodes are affected...

**Comment by bhalevy** [comment]
> @chripede did you manage to fix the issue by scrubing the sstables? I tried that with no luck.

@r4fek How did you try to scrub? Can you please paste the nodetool or api call you used and the node logs from scrub?

> @bhalevy is there any other way to fix this? Unofortunately all my nodes are affected...

Scrub would be my first choice, so let's verify that it ran properly and if there's an issue with it I'd like to understand and fix it.
The alternative without removing nodes would be to understand form the compaction logs which sstables contain the out-of-order partitions and see if we can remove them (after taking a snapshot), then let repair restore the data from other nodes.

With luck, the corruption is uncorrelated between the nodes and the data can be properly restored.

Scrub in SEGREGATE mode does delete out-of-order data but rather it sorts it into new sstables, allowing compaction and repair to proceed normally.

**Comment by r4fek** [comment]
Hmm and the weirdest thing is that these sstables even don't exist on the disk.

**Comment by bhalevy** [comment]
@r4fek the log messages you mentioned in https://github.com/scylladb/scylla/issues/9983#issuecomment-1037973432 aren't necessarily from scrub.  The latter should print messages like:
```
[Scrub compaction <ks.cf>] Invalid partition <key> (<key_hex>), partition is out-of-order compared to previous partition ...
```
when fixing out of order partitions.

> Hmm and the weirdest thing is that these sstables even don't exist on the disk.

Yes, the error is on the compaction write path, so it prevents it from writing corrupt sstables.
The question is whether the corruption already exists in its input sstables and if so, how did it get there.

@raphaelsc / @denesb, we need to rule out a bug in compaction.  If the input sstables are fine and we see the error only on the write path, it means that somehow regular compaction generates out-of-order tokens.

**Comment by bhalevy** [comment]
> @bhalevy in 4.5.3 sstables being validated / scrubbed are they separated away from regular compaction - or can compaction run on them and have them compacted to other sstables.

In 4.5 scrubbed sstables are separated from regular compaction, but there's a race there that was fixed only in 80a1ebf0f37642d3a64c8098308339d16f43730a.

That said, I don't think this race is relevant here.

**Comment by r4fek** [comment]
> @r4fek did you have in the past a mix SMP cluster - some nodes using one SMP count and some having a different count ?

I have 8 nodes with the same setup (number of cores, disks) since the beginning.
> 
> Can you please run https://docs.scylladb.com/troubleshooting/report-scylla-problem/#health-check-report and upload the info https://docs.scylladb.com/troubleshooting/report-scylla-problem/#send-files-to-scylladb-support

Yes of course. Report UUID: 8be79620-7400-4fbb-93e0-34dc5152505b
> 
> @bhalevy in 4.5.3 sstables being validated / scrubbed are they separated away from regular compaction - or can compaction run on them and have them compacted to other sstables.
> 
> @r4fek if you share a more complete log from a node that had those errors we can try to find out why we can locate the files that reported an error

scylla log also uploaded.

**Comment by bhalevy** [comment]
I suspect that the issue is related to having partition with many rows.
I'll try to reproduce locally.

@r4fek would it be possible to upload a few sstables that appear in the log that compaction fails to compact?
I can pass you their names privately.

**Comment by bhalevy** [comment]
Also, please confirm you're using LeveledCompactionStrategy.

@chripede what compaction strategy is it in your case?

**Comment by r4fek** [comment]
Yes, I'm using LeveledCompactionStrategy for some of the tables. Unfortunately I can't share them because they contains private users' data like passwords, sessions, bookmarks etc.

**Comment by chripede** [comment]
> strategy

@bhalevy We are using DateTieredCompactionStrategy

**Comment by bhalevy** [comment]
@chripede Can you please share the node logs during and after scrub?

Also, since I haven't been able to reproduce this locally yet, could you share a set of sstables that compaction can't compact together (and ideally all mentioned in the log so we can trace their history)?

**Comment by chripede** [comment]
> @chripede Can you please share the node logs during and after scrub?
> 
> Also, since I haven't been able to reproduce this locally yet, could you share a set of sstables that compaction can't compact together (and ideally all mentioned in the log so we can trace their history)?

@bhalevy Unfortunately I cannot. The node was decommissioned a week ago.
If there are any other details that could be of value, please let me know.

**Comment by bhalevy** [comment]
@r4fek did you upload any external sstables into the node via nodetool refresh?

**Comment by bhalevy** [comment]
@r4fek and what was roughly the scylla version history on your cluster?
The reason I'm asking is that I see sstables with no "origin" in their scylla metadata and those could have been created by an older version or uploaded via refresh.

**Comment by bhalevy** [comment]
For example, can you please search all the logs you have for the following sstables and/or see when they were written on the file system and correlate with the logs?

```
/var/lib/scylla/data/sync/password-fdc14e3094ac11e89e8000000000002b/md-1239946-big-Data.db
/var/lib/scylla/data/sync/password-fdc14e3094ac11e89e8000000000002b/md-1238981-big-Data.db
/var/lib/scylla/data/sync/password-fdc14e3094ac11e89e8000000000002b/md-1234663-big-Data.db
/var/lib/scylla/data/sync/password-fdc14e3094ac11e89e8000000000002b/md-1237153-big-Data.db
```

It appears like compaction fails to compact them and might already contain out-of-order partitions.

**Comment by r4fek** [comment]
@bhalevy 
Version history: 2.3 -> 2.3.1 -> 2.3.2 -> 3.0.2 -> 3.0.5 -> 3.0.7 -> 3.0.8 -> 3.0.10 -> 3.0.11 -> 3.1.2 -> 3.2.3 -> 4.0.1 -> 4.2.3 -> 4.4.1 -> 4.5.3

logs with required sstables are here:
* md-1239946-big-Data.db.log.tar.gz
* md-1238981-big-Data.db.log.tar.gz
* md-1234663-big-Data.db.log.tar.gz

Command used to fetch the logs: `journalctl -u scylla-server --since='2021-02-15 15:00' |grep /var/lib/scylla/data/sync/password-fdc14e3094ac11e89e8000000000002b/md-1238981-big-Data.db -C 5`

Can I help more?

**Comment by bhalevy** [comment]
> Can I help more?

Thanks much!

I have a few ideas in mind to help understand the problem but let me analyze the logs first and then we'll see.

Heads up, we've improved scrub significantly in 4.6 (which is close to be released by not available yet), including the SEGREGATE mode, so if indeed some on-disk sstables already contain out-of-order data (partitions and/or rows), we may need to wait for 4.6 to fix that, or backport the new scrub features to 4.5, but that might require similar time to do since 4.6 is imminent.

**Comment by r4fek** [comment]
I will update to 4.6 and see if it helps.

**Comment by bhalevy** [comment]
To summarize some intermediate findings from the per-sstable log:
```
lut 10 11:57:01 am4-dbs1 scylla[53268]:  [shard 31] compaction - [Compact sync.password 82471090-8a68-11ec-97c4-00000000001a] Compacted 1 sstables to [/var/lib/scylla/data/sync/password-fdc14e3094ac11e89e8000000000002b/md-1234663-big-Data.db:level=0, ]. 491MB to 491MB (~100% of original) in 21008ms = 23MB/s. ~15488 total partitions merged to 15364.

lut 11 17:40:54 am4-dbs1 scylla[28828]:  [shard 31] compaction - [Compact sync.password c3078020-8b61-11ec-b790-2fcf91f03416] Compacting [/var/lib/scylla/data/sync/password-fdc14e3094ac11e89e8000000000002b/md-1234663-big-Data.db:level=0:origin=, ]

lut 11 17:40:55 am4-dbs1 scylla[28828]:  [shard 31] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x61f00ad0eb00 for sstable writer /var/lib/scylla/data/sync/password-fdc14e3094ac11e89e8000000000002b/md-1242615-big-Data.db (sync.password fdc14e30-94ac-11e8-9e80-00000000002b)] Unexpected token: previous -8185000295678365556, current -8185015509266174075): retrying
```

Tells me that `md-1234663-big-Data.db` was created by compaction in a pre-4.5 version (assuming 4.4.1) that doesn't support the sstable origin yet.

Writing this sstable passed the validator checks.

Re-compacting the same sstable, on its own, trips the validation error for `Unexpected token`.

So the possible options are that either:
1. compaction managed to generate a corrupt sstable, somehow escaping the validator checks the first time around, or
2. re-compacting the sstable produces an invalid stream that the write-validator rejects.

The latter may implicate the data, the index, or both.

**Comment by bhalevy** [comment]
@r4fek when exactly did you upgrade from 4.4.1 to 4.5.3?

It looks like `md-1234663-big-Data.db` was written using scylla-4.4.1
and later attempted compaction by 4.5.3.

Looking at the 4.4 code, there was no validation of keys on the write path by default.

This was changed in 4.5 (ef97adc72a0d5e8877885810b1a7e0fcbb11dc29)
so it's possible that the sstable in question was written as out-of-order keys (that compaction could have carried over from its inputs files) after all.

**Comment by r4fek** [comment]
I upgraded Scylla to 4.5.3 3 days ago and I'm quite sure these errors were visible on 4.4.1 too.

**Comment by raphaelsc** [comment]
> I upgraded Scylla to 4.5.3 3 days ago and I'm quite sure these errors were visible on 4.4.1 too.

do you set Scylla's option enable_sstable_key_validation to true?

So you moved from 4.4.1 on Feb 13th right? According to the log shared by you, the validator started yelling at Feb 11th.

**Comment by r4fek** [comment]
I double checked and I moved from 4.4.1 to 4.5.3 on Feb 10th, sorry for confusion! It means that validator started yelling right after the upgrade to 4.5.3:

```
$ journalctl -u scylla-server --since="2021-03-03" |grep invalid_mutation_fragment_stream |head
lut 10 17:18:20 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d009bb5b00 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44534661-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
lut 10 17:18:35 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d008f0cb00 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44534829-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
lut 10 17:18:53 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d008f06300 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44534997-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
lut 10 17:19:17 am4-dbs1 scylla[28828]:  [shard 42] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62a0085dc300 for sstable writer /var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/md-1056762-big-Data.db (sync.session_by_id 051e3940-94ad-11e8-93cf-000000000017)] Unexpected token: previous -4261495074737813676, current -4261527946879923112): retrying
lut 10 17:19:22 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d009bb8300 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44535165-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
lut 10 17:19:57 am4-dbs1 scylla[28828]:  [shard 42] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62a00ab46300 for sstable writer /var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/md-1057434-big-Data.db (sync.session_by_id 051e3940-94ad-11e8-93cf-000000000017)] Unexpected token: previous -4261495074737813676, current -4261527946879923112): retrying
lut 10 17:20:10 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d008b91b00 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44535333-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
lut 10 17:20:38 am4-dbs1 scylla[28828]:  [shard 42] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62a008dd0b00 for sstable writer /var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/md-1057938-big-Data.db (sync.session_by_id 051e3940-94ad-11e8-93cf-000000000017)] Unexpected token: previous -4261495074737813676, current -4261527946879923112): retrying
lut 10 17:21:24 am4-dbs1 scylla[28828]:  [shard 42] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62a008dd0b00 for sstable writer /var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/md-1058442-big-Data.db (sync.session_by_id 051e3940-94ad-11e8-93cf-000000000017)] Unexpected token: previous -4261495074737813676, current -4261527946879923112): retrying
lut 10 17:21:38 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d00881c300 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44535501-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
```

**Comment by r4fek** [comment]
Upgrade was performed on `Thu 2022-02-10 ~17-18 UTC; 6 days ago`

**Comment by raphaelsc** [comment]
md-1234663-big-Data.db belongs to shard 31, could you please upload logs from shard 31 since Jan 1th (by grepping "shard 31"; removing lots of unnecessary noise)? I'd like to trace it back into its origin.

**Comment by raphaelsc** [comment]
> I double checked and I moved from 4.4.1 to 4.5.3 on Feb 10th, sorry for confusion! It means that validator started yelling right after the upgrade to 4.5.3:
> 
> ```
> $ journalctl -u scylla-server --since="2021-03-03" |grep invalid_mutation_fragment_stream |head
> lut 10 17:18:20 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d009bb5b00 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44534661-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
> lut 10 17:18:35 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d008f0cb00 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44534829-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
> lut 10 17:18:53 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d008f06300 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44534997-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
> lut 10 17:19:17 am4-dbs1 scylla[28828]:  [shard 42] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62a0085dc300 for sstable writer /var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/md-1056762-big-Data.db (sync.session_by_id 051e3940-94ad-11e8-93cf-000000000017)] Unexpected token: previous -4261495074737813676, current -4261527946879923112): retrying
> lut 10 17:19:22 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d009bb8300 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44535165-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
> lut 10 17:19:57 am4-dbs1 scylla[28828]:  [shard 42] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62a00ab46300 for sstable writer /var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/md-1057434-big-Data.db (sync.session_by_id 051e3940-94ad-11e8-93cf-000000000017)] Unexpected token: previous -4261495074737813676, current -4261527946879923112): retrying
> lut 10 17:20:10 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d008b91b00 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44535333-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
> lut 10 17:20:38 am4-dbs1 scylla[28828]:  [shard 42] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62a008dd0b00 for sstable writer /var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/md-1057938-big-Data.db (sync.session_by_id 051e3940-94ad-11e8-93cf-000000000017)] Unexpected token: previous -4261495074737813676, current -4261527946879923112): retrying
> lut 10 17:21:24 am4-dbs1 scylla[28828]:  [shard 42] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62a008dd0b00 for sstable writer /var/lib/scylla/data/sync/session_by_id-051e394094ad11e893cf000000000017/md-1058442-big-Data.db (sync.session_by_id 051e3940-94ad-11e8-93cf-000000000017)] Unexpected token: previous -4261495074737813676, current -4261527946879923112): retrying
> lut 10 17:21:38 am4-dbs1 scylla[28828]:  [shard 45] compaction_manager - compaction failed: invalid_mutation_fragment_stream ([validator 0x62d00881c300 for sstable writer /var/lib/scylla/data/sync/session-0384385094ad11e88e02000000000025/md-44535501-big-Data.db (sync.session 03843850-94ad-11e8-8e02-000000000025)] Unexpected token: previous -8953998140184991728, current -8954013037616868992): retrying
> ```

No problem. Good to know.

**Comment by r4fek** [comment]
I'd like to upload shard31 log, but looks like you have some problems with uploading reports right now:
```
curl -X PUT https://upload.scylladb.com/$report_uuid/shard31.log.tar.gz -T shard31.log.tar.gz
<?xml version='1.0' encoding='UTF-8'?><Error><Code>AccessDenied</Code><Message>Access denied.</Message><Details>Anonymous caller does not have storage.objects.delete access to the Google Cloud Storage object
```

**Comment by r4fek** [comment]
> do you set Scylla's option enable_sstable_key_validation to true?

Nope. Nothing changed in my settings.

**Comment by denesb** [comment]
> > I upgraded Scylla to 4.5.3 3 days ago and I'm quite sure these errors were visible on 4.4.1 too.
> 
> do you set Scylla's option enable_sstable_key_validation to true?
> 

The reason 4.5 started yelling and why 4.4 didn't is that 4.5 has ef97adc72a0d5e8877885810b1a7e0fcbb11dc29 which enabled unconditional token order validation. This was deemed cheap enough to do unconditionally on all sstable write paths. Previously keys were not at all validated by default.

**Comment by raphaelsc** [comment]
> > > I upgraded Scylla to 4.5.3 3 days ago and I'm quite sure these errors were visible on 4.4.1 too.
> > 
> > do you set Scylla's option enable_sstable_key_validation to true?
> > 
> 
> The reason 4.5 started yelling and why 4.4 didn't is that 4.5 has ef97adc72a0d5e8877885810b1a7e0fcbb11dc29 which enabled unconditional token order validation. This was deemed cheap enough to do unconditionally on all sstable write paths. Previously keys were not at all validated by default.
> 
> 

Yes. The reason I asked that was to understand how come he saw validation errors on 4.4.1 which doesn't have token validation enabled by default, the only reasonable explanation would be that option explicitly enabled, but it turned out that errors only showed up after upgrade to 4.5, which is what we really expected.

**Comment by bhalevy** [comment]
> Yes. The reason I asked that was to understand how come he saw validation errors on 4.4.1 which doesn't have token validation enabled by default, the only reasonable explanation would be that option explicitly enabled, but it turned out that errors only showed up after upgrade to 4.5, which is what we really expected.

@raphaelsc it is possible there were other errors reported with 4.4.1, at the mutation_fragment kind level (e.g. missing `partition_end`)

**Comment by denesb** [comment]
After some further investigation this looks like https://github.com/scylladb/scylla/issues/7552.
Note that despite the cluster not having heterogeneous shard counts (neither currently nor in the past) it is still possible to hit the RPC reordering bug. RPC connections are not guaranteed to land on the same shard on the remote node. It is very likely but not guaranteed. So it is possible that even in a cluster with the same number of shards on all nodes, some RPC messaging involves `smp::submit_to()` which could have reordered some of those messages.
The reordered data was most likely generated back when the cluster was still on a vulnerable version, and its existence was uncovered by upgrading to a version which has the token validation on the write path.
The fix is to upgrade to 4.6 (once it comes out) and scrub with SEGREGATE mode.

**Comment by bhalevy** [comment]
With no further information, closing as duplicate of https://github.com/scylladb/scylla/issues/7552
